rfcs/9999-new-merkle-storage.md

```diff
@@ -0,0 +1,140 @@
+There is a time-memory trade off here - hashing allows deduplication but takes
+time. If we didn't want deduplication we could defer hashing until we actually
+need the hash. But we do, hence the GC driven compromise. The queue must also be
+drained whenever we look up an object by hash.
+
+### Public key -> Merkle address mapping
+
+There is a problem with this scheme however. Because our ledger is unordered,
+you need the mapping of public keys to addresses in the tree in order to
+efficiently read from it. So a `Ref` to the ledger alone is insufficient, we
+need to store additional data alongside it. We could keep it separate, but we
+want it persisted to disk, which would mean more database handling logic. So I
+- reluctantly - think it's best to Merkleize the mapping as some binary search
+tree and store that using the new system. The ledger would then be a combination
+of the mapping and the complete binary tree we currently use. We'll continue to
+only SNARK the non-indexed tree.
+
+## Drawbacks
+[drawbacks]: #drawbacks
+
+This is a lot of engineering work. It'll be more complex than any individual
+component of the current system, though less complex than all of them taken
+together. 
+
+## Rationale and alternatives
+[rationale-and-alternatives]: #rationale-and-alternatives
+
+* Why is this design the best in the space of possible designs?
+  The alternative is continuing with the current system and continuing to extend
+  it. It's possible to make the current system a bit more principled and less
+  error prone, but it'll always be a drag on productivity. Masks are not really
+  the "right" model for what we're talking about.
+* What other designs have been considered and what is the rationale for not choosing them?
+  A much gentler revision to how masks are handled is possible: remove the
+  `unregister_mask_exn` function and use a finalizer to trigger that behavior.
+  That would remove the possibility of unregistering them in the wrong order or
+  forgetting to do so. It'd be an improvement, but a much smaller one than
+  switching to this new system.
+* What is the impact of not doing this?
+  Continued pain working with the current system, especially as we add more
+  stuff to it.
+
+## Prior art
+[prior-art]: #prior-art
+
+The conceptual model here, of a generic Merkle DAG, is pretty much stolen from
+IPFS. If IPFS itself weren't so heavy I'd consider just using it for the storage
+system, but it is, and there are other issues with that.
+
+## Unresolved questions
+[unresolved-questions]: #unresolved-questions
+
+* What parts of the design do you expect to resolve through the RFC process before this gets merged?
+ * I hate `Ref` as a name for the type. Can somebody suggest a better one?
+ * How does this interact with versioning? We'd want to version both the RPCs
+   for the new sync code and the data being synced.
+ * How do changes to the OCaml Merkle system affect Snarky code?
```

Comment:
For the most part not at all. The SNARK requests are

```ocaml
  type path = Pedersen.Digest.t list

  type _ Request.t +=
    | Get_path : Account.Index.t -> path Request.t
    | Get_element : Account.Index.t -> (Account.t * path) Request.t
    | Set : Account.Index.t * Account.t -> unit Request.t
    | Find_index : Public_key.Compressed.t -> Account.Index.t Request.t
```

Pretty easy to implement with this system.